collie/model/hybrid_matrix_factorization.py

```diff
@@ -207,38 +239,102 @@ def __init__(self,
                         'stage': 'matrix_factorization',
                     },
                 ]
-
-            optimizer_config_list = initial_optimizer_block + [
-                {
-                    'lr': metadata_only_stage_lr,
-                    'optimizer': metadata_only_stage_optimizer,
-                    # optimize metadata layers only
-                    'parameter_prefix_list': ['metadata', 'combined', 'user_bias', 'item_bias'],
-                    'stage': 'metadata_only',
-                },
-                {
-                    'lr': all_stage_lr,
-                    'optimizer': all_stage_optimizer,
-                    # optimize everything
-                    'parameter_prefix_list': ['user', 'item', 'metadata', 'combined'],
-                    'stage': 'all',
-                },
-            ]
-
+            if item_metadata is not None and user_metadata is not None:
+                optimizer_config_list = initial_optimizer_block + [
+                    {
+                        'lr': metadata_only_stage_lr,
+                        'optimizer': metadata_only_stage_optimizer,
+                        # optimize metadata layers only
+                        'parameter_prefix_list': [
+                            'item_metadata', 'combined', 'user_bias', 'item_bias'
+                        ],
+                        'stage': 'item_metadata_only',
+                    },
+                    {
+                        'lr': metadata_only_stage_lr,
+                        'optimizer': metadata_only_stage_optimizer,
+                        # optimize metadata layers only
+                        'parameter_prefix_list': [
+                            'user_metadata', 'combined', 'user_bias', 'item_bias'
+                        ],
+                        'stage': 'user_metadata_only',
+                    },
```

Comment:
Looking at this all now, I think I might be leaning more towards combining these two stages into a single metadata one. If we don't, my worry is that the item metadata stage's output will be reliant on a frozen user metadata output that is essentially just garbage output, so the item output will have to compensate a lot to improve performance. 

My thought is that if they are both frozen / unfrozen at the same time, the model can properly learn user and item features together where it makes sense. 

What do you think?

Reply:
I am torn tbh. As I understand it, by combining them into a single metadata layer we make the hybrid model more similar to the pre-trained, which is a good thing. 

However, it feels valuable to be able to look at model performance at the `item_metadata_only` stage and then compare it to what happens to the metrics after you bring in the `user_metadata` in the next stage. 

Now for the however to my however, heh. We can technically still do this by creating multiple separate models with/without the different metadata. This is slightly more work for the user, but they do get to compare the models at their final stage, which is probably a better way to do it and comparing at stages 2 and 3 may be something that we want to discourage?  

Ok, after all of that I think I am going to go ahead and make these changes so that it's a single stage. We can always revert later.

Reply:
I feel pretty split about it too. I feel like between the two options, having it all be a single stage is slightly easier to understand / will lead to better default performance for most cases. I think for advanced users who might want that granularity, it should still be possible to make a subclass changing this optimizer setup to get that done.